### ISSUE TITLE
Run Tests Sequentially 

### ISSUE THREAD

**Issue by lasuk** [macxred/cashctrl_ledger#13]
**Description:**

Currently, our integration tests are set up to run in parallel within GitHub Actions. These tests connect to a remote test server via REST API and include checks to ensure mutator methods affect the remote test account as expected. However, these tests cannot run concurrently because they access and modify the same remote instance, leading to conflicts and unreliable test results. This issue typically arises when pushing changes to a PR, since tests are triggered both for the feature branch and the PR. We need to update our GitHub Actions configuration to run these tests sequentially instead of in parallel to ensure reliable and consistent testing outcomes.

Additionally, another project (`cashctrl_api`) connects to the same test instance. We need to either ensure sequential execution of tests between both projects, assign a different test account to the other project, or accept false negatives in the unlikely case of concurrently running tests. We should adapt the daily scheduled test times to avoid concurrent scheduled tests.

Furthermore, we should automatically reset the test environment to the original state every day by providing a script to restore the original state on the test server and executing it automatically once a day.

**Tasks:**

1. **Modify GitHub Actions Workflow:**
   - Update the `.github/workflows` configuration file to ensure that tests are run sequentially.
   - Ensure that the entire test suite waits for one test to complete before starting the next.
   - Coordinate with the `cashctrl_api` project to ensure tests do not run concurrently, either by sequential execution, assigning a different test account, or accepting the risk of false negatives.

2. **Adapt Scheduled Test Times:**
   - Adjust the daily scheduled test times to prevent concurrent test execution between both projects.
   - Ensure the schedule allows adequate time for each project's tests to complete before the other project's tests begin.

3. **Provide and Execute Restoration Script:**
   - Develop a script to restore the original state of the test server.
   - Integrate this script into the CI pipeline and schedule it to run automatically once a day to reset the test environment.

**Resources:**

A quick google search brought up these links. Use them if helpful, ignore otherwise:
- https://docs.github.com/en/actions/using-jobs/using-concurrency
- https://stackoverflow.com/questions/78315405/github-actions-sequential-merge-queue

**Comment by AlexTheWizardL** [comment]
Also, need to check if one test creates data that is used in the next one. For example, one test creates VAT code and assign it to some account and then another test can not delete it

**Comment by lasuk** [comment]
From https://docs.github.com/en/actions/using-jobs/using-concurrency:

> The default behavior of GitHub Actions is to allow multiple jobs or workflow runs to run concurrently. The concurrency keyword allows you to control the concurrency of workflow runs.
> [..]
> The concurrency key is used to group workflows or jobs together into a concurrency group. When you define a concurrency key, GitHub Actions ensures that only one workflow or job with that key runs at any given time. If a new workflow run or job starts with the same concurrency key, GitHub Actions will cancel any workflow or job already running with that key. The concurrency key can be a hard-coded string, or it can be a dynamic expression that includes context variables.
> [..]
> This means that when a workflow run or job starts, GitHub will cancel any workflow runs or jobs that are already in progress in the same concurrency group. This is useful in scenarios where you want to prevent parallel runs for a certain set of a workflows or jobs, such as the ones used for deployments to a staging environment, in order to prevent actions that could cause conflicts or consume more resources than necessary.

**Comment by lasuk** [comment]
Opened a separate issue #23 for to automatically reset the test environment to the original state.

**Comment by lasuk** [comment]
Implemented in #22
